### PR TITLE
First

### PR DESCRIPTION
**Pull request by mujtaba-saboor**

### CODE REVIEW

File: assignment.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
+require_relative "event_record.rb"
+event = {add: "1", month_view: "2", date_view: "3", all_events: "4", delete: "5", exit: "6"}
```
Comment (kashif-umair) [inline assignment.rb]: use rubocop https://github.com/rubocop-hq/rubocop

File: assignment.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
+require_relative "event_record.rb"
+event = {add: "1", month_view: "2", date_view: "3", all_events: "4", delete: "5", exit: "6"}  
+loop do 
+  puts
+  puts "For Adding an event, Enter: 1"
```
Comment (kashif-umair) [inline assignment.rb]: move numbers at the start of lines.

File: assignment.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  line = gets.strip
+  events = []
+  count = -1
+  case line
+    when event[:add]
+      puts
+      check = "y"
+      while (check == "y")
+        count += 1
+        events << EventRecord.add_event
```
Comment (kashif-umair) [inline assignment.rb]: - no need to wrap .new in a method.
- pass details while calling `.new` method.

File: assignment.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  case line
+    when event[:add]
+      puts
+      check = "y"
+      while (check == "y")
+        count += 1
+        events << EventRecord.add_event  
+        puts "Enter name of the Event"
+        name = gets.strip
+        puts "Enter date for the event (yyyy:mm:dd)"
```
Comment (kashif-umair) [inline assignment.rb]: wrong date format mentioned.

File: assignment.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        date = gets.strip
+        events[count].add_event_details(name, date)
```
Comment (kashif-umair) [inline assignment.rb]: you don't need this count variable. push an object into array only when it's final.

File: assignment.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        puts "Do u wish to add another event(y/n)?"
+        check = gets.strip
```
Comment (kashif-umair) [inline assignment.rb]: downcase

File: assignment.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+      check = "y"
+      while (check == "y")
+        count += 1
+        events << EventRecord.add_event  
+        puts "Enter name of the Event"
+        name = gets.strip
+        puts "Enter date for the event (yyyy:mm:dd)"
+        date = gets.strip
+        events[count].add_event_details(name, date)
+        puts "Do u wish to add another event(y/n)?"
```
Comment (kashif-umair) [inline assignment.rb]: u -> you

File: assignment.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        check = gets.strip
+      end
+    when event[:month_view]
+      puts
+      EventRecord.events_for_given_month  
+    when event[:date_view]
+      puts
```
Comment (kashif-umair) [inline assignment.rb]: try to follow DRY.

File: assignment.rb [hunk 9]
```diff
@@ -0,0 +1,47 @@
+require_relative "event_record.rb"
+event = {add: "1", month_view: "2", date_view: "3", all_events: "4", delete: "5", exit: "6"}  
+loop do 
+  puts
+  puts "For Adding an event, Enter: 1"
+  puts "For Viewing a specific month's event(s), Enter: 2"
+  puts "For Viewing a specific date's event(s), Enter: 3"
+  puts "For Viewing all events, Enter: 4"
+  puts "For Deleting an event, Enter: 5"
+  puts "To Quit, Enter 6"
```
Comment (kashif-umair) [inline assignment.rb]: move everything to methods and classes to use pure OOP approach.

File: event_record.rb [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+class EventRecord
+  @@events_date = Hash.new{|hash, key| hash[key] = []}
+  @@events_details = {}
+  @@id = 0
+
+  def add_event_details(name, date)
+    begin
+      date = Date.parse(date)
+      @@id += 1
+      @@events_details[@@id] = Event_Details.new(@@id, name, date)
```
Comment (kashif-umair) [inline event_record.rb]: - class name violates ruby naming convention.
- Save only the date against ID. Avoid creating duplicate events with same information.

File: event_record.rb [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+      year_month_key = "#{date.year.to_s}-#{date.month.to_s}"
+      day = date.day
+      @@events_date[year_month_key].push([name, day])
+      puts "Event added"
+      rescue
```
Comment (kashif-umair) [inline event_record.rb]: - indentation.
- show proper message to user in case of exception.

File: event_record.rb [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        puts "Incorrect Data"
+    end
+  end
+
+  def self.return_all_events
+    if @@events_details.length == 0
```
Comment (kashif-umair) [inline event_record.rb]: write success scenario first and then blank scenario.

File: event_record.rb [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+      puts "No event added yet"
+    else
+      @@events_details.each do |key, event_detail|
+        puts "ID: #{event_detail.id} Name: #{event_detail.event_name.ljust(7)} Date: #{event_detail.date.strftime("%Y/%m/%d")}"
```
Comment (kashif-umair) [inline event_record.rb]: `puts event_detail`

File: event_record.rb [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+    end
+    end
```
Comment (kashif-umair) [inline event_record.rb]: indentation.

File: event_record.rb [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+  def self.events_for_given_month
+    if @@events_details.length == 0
+      @@id = 0
+      puts "No event added yet"
+    else
+      begin
+        puts "Enter the required year (yyyy)"
+        year = gets.strip.to_i
+        puts "Enter the required month (mm)"
+        month = gets.strip.to_i
```
Comment (kashif-umair) [inline event_record.rb]: get the input in one go.

File: event_record.rb [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+        date = DateTime.new(year, month)
+        year_month_key = "#{date.year.to_s}-#{date.month.to_s}"
```
Comment (kashif-umair) [inline event_record.rb]: remove `.to_s`

File: event_record.rb [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+        if [1, 3, 5, 7, 8, 10, 12].include?(date.month)
+          range = (1..31)
+        elsif [4, 6, 9, 11].include?(date.month)
+          range = (1..30)
+        elsif Date.leap?(date.year)
+          range = (1..29)
```
Comment (kashif-umair) [inline event_record.rb]: use built-in logic for this.

File: event_record.rb [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+        else
+          range = (1..28)
+        end
+        month_event_detail_hash = Hash.new{|hash, key| hash[key] = []}
+        if @@events_date.has_key? (year_month_key)
+          @@events_date[year_month_key].each do |name_day|
+            month_event_detail_hash[name_day[1].to_i] << name_day[0]
+          end
+        puts "Mon Tue Wed Thu Fri Sat Sun"
```
Comment (kashif-umair) [inline event_record.rb]: use tabs.

File: event_record.rb [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+        dates = [nil] * 2 + (1..range.last).to_a
```
Comment (kashif-umair) [inline event_record.rb]: this logic is not right.

File: event_record.rb [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+        puts "Enter the required date (yyyy-mm-dd)"
+        date = Date.parse(gets.strip)
+        year_month_key = "#{date.year.to_s}-#{date.month.to_s}"
+        month_event_detail_hash = Hash.new{ |hash, key| hash[key] = [] }
+        if @@events_date.has_key? (year_month_key)
+          @@events_date[year_month_key].each do |name_day|
+            month_event_detail_hash[name_day[1].to_i] << name_day[0]
+          end
+          if month_event_detail_hash.has_key? (date.day)
+            puts "Event(s) Scheduled: #{month_event_detail_hash[date.day]}"
```
Comment (kashif-umair) [inline event_record.rb]: Events Scheduled:

Hello
World
Foo
Bar

File: event_record.rb [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+    end
+  end
+  
+  def self.delete_event
+    if @@events_details.length == 0
+      puts "No event added yet"
+      @@id = 0
+    else
+      puts "Enter an Id from the list of events given below"
+      self.return_all_events
```
Comment (kashif-umair) [inline event_record.rb]: `self.` can be removed.

**Comment by kashif-umair** [comment]
Cycle 1

**Comment by mujtaba-saboor** [comment]
Resolved the 159 issues identified by Rubocop down to 24, 
the remaining issues are mentioned below:
- Assignment Branch Condition size for events_for_given_date is too high. [26.98/15]
- Use Hash#key? instead of Hash#has_key?.
- Replace class var @@id with a class instance var. (multiple)
- Perceived complexity for events_for_given_month is too high. [12/7]
- Method has too many lines. [48/10] def self.events_for_given_month ...
- Cyclomatic complexity for events_for_given_month is too high. [8/6]
-  Missing top-level class documentation comment.

File: event_record.rb (hunk 10 above)
Comment (mujtaba-saboor) [reply event_record.rb]: done

File: assignment.rb (hunk 7 above)
Comment (mujtaba-saboor) [reply assignment.rb]: done

File: assignment.rb (hunk 9 above)
Comment (mujtaba-saboor) [reply assignment.rb]: done

File: assignment.rb (hunk 6 above)
Comment (mujtaba-saboor) [reply assignment.rb]: done

File: assignment.rb (hunk 5 above)
Comment (mujtaba-saboor) [reply assignment.rb]: done

File: assignment.rb (hunk 4 above)
Comment (mujtaba-saboor) [reply assignment.rb]: done

File: assignment.rb (hunk 3 above)
Comment (mujtaba-saboor) [reply assignment.rb]: done

File: assignment.rb (hunk 1 above)
Comment (mujtaba-saboor) [reply assignment.rb]: done

File: assignment.rb (hunk 2 above)
Comment (mujtaba-saboor) [reply assignment.rb]: done

File: assignment.rb (hunk 8 above)
Comment (mujtaba-saboor) [reply assignment.rb]: done

File: event_record.rb (hunk 11 above)
Comment (mujtaba-saboor) [reply event_record.rb]: done

File: event_record.rb (hunk 12 above)
Comment (mujtaba-saboor) [reply event_record.rb]: done

File: event_record.rb (hunk 13 above)
Comment (mujtaba-saboor) [reply event_record.rb]: done

File: event_record.rb (hunk 14 above)
Comment (mujtaba-saboor) [reply event_record.rb]: done

File: event_record.rb (hunk 16 above)
Comment (mujtaba-saboor) [reply event_record.rb]: done

File: event_record.rb (hunk 19 above)
Comment (mujtaba-saboor) [reply event_record.rb]: done

File: event_record.rb (hunk 21 above)
Comment (mujtaba-saboor) [reply event_record.rb]: done

File: event_record.rb (hunk 20 above)
Comment (mujtaba-saboor) [reply event_record.rb]: done

File: assignment.rb (hunk 3 above)
Comment (mujtaba-saboor) [reply assignment.rb]: done

File: event_record.rb (hunk 15 above)
Comment (mujtaba-saboor) [reply event_record.rb]: done

File: event_record.rb (hunk 17 above)
Comment (mujtaba-saboor) [reply event_record.rb]: done

File: event_record.rb (hunk 18 above)
Comment (mujtaba-saboor) [reply event_record.rb]: done

**Pull request closed without merging**